### PR TITLE
feat: Add admin API endpoints for runtime config reload

### PR DESCRIPTION
**Pull request by theTibi**
### Problem

When HolmesGPT's configuration needs to change (toolsets, models, runbooks), the only option today is to restart the entire server process. This causes downtime, drops in-flight requests, and is slow in environments where the container takes time to initialize (loading toolsets, connecting to databases, etc.).

This is especially painful in production Kubernetes/Docker deployments where config files are mounted via ConfigMaps or bind mounts and updated frequently.

### Changes

- **`holmes/config.py`**: Add `reload_toolsets()` and `reload_models()` methods with thread-safe locking. Add lock protection to `llm_model_registry` lazy property.
- **`holmes/admin/admin_api.py`** *(new)*: FastAPI sub-app mounted at `/api/admin` with three POST endpoints:
  - `POST /api/admin/reload/toolsets` — re-reads config YAML, rebuilds toolsets/runbooks
  - `POST /api/admin/reload/models` — re-reads `model_list.yaml`, rebuilds model registry
  - `POST /api/admin/reload` — reloads everything
- **`server.py`**: Mount admin sub-app (2 lines)
- **`tests/test_config_reload.py`** *(new)*: Unit + integration tests

### Example response

```json
{
  "status": "ok",
  "component": "all",
  "detail": "50 toolsets (15 enabled), 4 models",
  "counts": {
    "toolsets_total": 50,
    "toolsets_enabled": 15,
    "runbooks": 3,
    "models_loaded": 4
  }
}
```

### Test plan

- [x] Unit tests for `reload_toolsets()` (reset state, YAML pickup, no-config edge case)
- [x] Unit tests for `reload_models()` (registry reset, model count)
- [x] Integration tests for all 3 endpoints (200 happy path, 500 error path)
- [x] Manual testing against live Docker deployment with bind-mounted config

### Notes

- Admin endpoints are unauthenticated, matching existing API patterns. 
- Reload is safe during active requests — in-flight requests use old config, new requests pick up reloaded config.

## Summary by CodeRabbit

* **New Features**
  * Added three unauthenticated admin POST endpoints: /api/admin/reload (all), /api/admin/reload/toolsets (toolsets/runbooks), and /api/admin/reload/models (model registry). Each returns operation counts and human-readable details; failures include error detail. Endpoints are only active when the admin API is enabled in server config.

* **Documentation**
  * API reference updated with endpoint details, example 500 payload for model reload, and a note advising network-level access restriction.

* **Tests**
  * Added tests for reload operations, endpoint responses, and error propagation.

### CODE REVIEW

File: holmes/admin/admin_api.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    detail: str = ""
+    counts: Dict[str, int] = {}
+
+
+def init_admin_app(main_app: FastAPI, config: Config, dal: SupabaseDal):
+    # TODO: Add authentication (API key header or similar) before exposing in production.
+    global _CONFIG, _DAL
+    _CONFIG = config
+    _DAL = dal
+    main_app.mount("/api/admin", admin_app)
```
[the remark this answers is not in the document]
Comment (theTibi) [reply holmes/admin/admin_api.py]: https://github.com/HolmesGPT/holmesgpt/pull/1948 will address that issue.

File: holmes/config.py [hunk 2]
```diff
@@ -434,6 +435,42 @@ def refresh_tool_executor(
...
+            if fresh is not None:
+                self.toolsets = fresh.toolsets
+                self.mcp_servers = fresh.mcp_servers
+                self.custom_toolsets = fresh.custom_toolsets
+                self.custom_runbook_catalogs = fresh.custom_runbook_catalogs
+            self._toolset_manager = None
+            self._cached_tool_executor = None
+            self._cached_executor_key = None
+        logging.info("Toolset config reloaded from %s", self._config_file_path)
+        return {"reloaded": True}
```
[the remark this answers is not in the document]
Comment (theTibi) [reply holmes/config.py:540]: it was fixed, review it again

File: holmes/admin/admin_api.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        counts["models_loaded"] = model_result.get("models_loaded", 0)
+        return ReloadResponse(
+            status="ok",
+            component="all",
+            detail=f"{counts['toolsets_total']} toolsets ({counts['toolsets_enabled']} enabled), {counts['models_loaded']} models",
+            counts=counts,
+        )
+    except Exception as e:
+        logging.error("Failed to reload all config", exc_info=True)
+        raise HTTPException(status_code=500, detail=str(e)) from e
```
[the remark this answers is not in the document]
Comment (theTibi) [reply holmes/admin/admin_api.py:119]: This is intentional. Both reload_toolsets() and reload_models() already acquire _executor_lock internally, so wrapping them with the same lock from outside would deadlock (it's a threading.Lock, not RLock). 

Even if we switched to RLock, the practical benefit is negligible: the window between the two reloads is microseconds, the admin reload is an explicit operator action (not automated), and an in-flight /api/chat seeing toolsets-refreshed-but-models-stale for a single request is harmless since both are independently valid states.

File: holmes/config.py [hunk 4]
```diff
@@ -159,9 +159,10 @@ def dal(self) -> SupabaseDal:
 
     @property
     def llm_model_registry(self) -> LLMModelRegistry:
-        if not self._llm_model_registry:
-            self._llm_model_registry = LLMModelRegistry(self, dal=self.dal)
-        return self._llm_model_registry
+        with self._executor_lock:
+            if not self._llm_model_registry:
+                self._llm_model_registry = LLMModelRegistry(self, dal=self.dal)
+            return self._llm_model_registry
```
[the remark this answers is not in the document]
Comment (theTibi) [reply holmes/config.py:203]: The registry is constructed once (at startup or after an explicit admin reload) and reads a small local YAML file -- the lock hold time is negligible. Adding a second lock would complicate the reload path (which already uses _executor_lock to null out the registry) and increase the risk of lock-ordering bugs for no practical benefit. Before this PR, there was no lock here at all; _executor_lock is a net improvement.

**Comment by moshemorad** [comment]
@claude review

File: holmes/config.py [hunk 5]
```diff
@@ -451,6 +452,66 @@ def refresh_tool_executor(
...
+        if self._config_file_path and Path(self._config_file_path).exists():
+            fresh = load_model_from_file(Config, Path(self._config_file_path))
+
+        with self._executor_lock:
+            if fresh is not None:
+                self.toolsets = fresh.toolsets
+                self.mcp_servers = fresh.mcp_servers
+                self.custom_toolsets = fresh.custom_toolsets
+                self.custom_runbook_catalogs = fresh.custom_runbook_catalogs
+                self.additional_toolsets = fresh.additional_toolsets
```
[the remark this answers is not in the document]
Comment (theTibi) [reply holmes/config.py:528]: @claude can you recheck this comment ?

**Comment by theTibi** [comment]
@moshemorad can you trigger another review on this, I have updated the branch and renamed runbooks to skills.

**Comment by moshemorad** [comment]
@claude review

File: holmes/config.py [hunk 6]
```diff
@@ -498,6 +499,67 @@ def refresh_tool_executor(
...
+        if self._config_file_path and Path(self._config_file_path).exists():
+            fresh = load_model_from_file(Config, Path(self._config_file_path))
+
+        with self._executor_lock:
+            if fresh is not None:
+                self.model = fresh.model
+                self.api_key = fresh.api_key
+                self.api_base = fresh.api_base
+                self.api_version = fresh.api_version
+                self.fast_model = fresh.fast_model
```
[the remark this answers is not in the document]
Comment (theTibi) [reply holmes/config.py:559]: @moshemorad please trigger another review

**Comment by moshemorad** [comment]
@claude review

**Review by moshemorad — APPROVED** [review]
